File: labview/src/analyze/graph.ts
import type { AppStack, Service, Graph, GraphNode, GraphEdge } from "../model/types.js";
import { routerIsServing } from "../labels/auth.js";
import { primaryIngress } from "../model/ingress.js";
import { graphServiceId } from "../model/networks.js";
import type { ResolvedDependency } from "./dependencies.js";
import { buildNetworkIndex, serviceKey, sharedWith, type NetworkIndex } from "./networks.js";

/**
 * Build the relationship graph across all stacks:
 *  - service <-> network membership (real docker network names, so shared/external
 *    networks correctly connect services from different stacks), each network node
 *    carrying its scope and how many services and stacks it joins
 *  - service -> service dependency, drawn **through** the network that carries it:
 *    the membership edges either side of the shared network take the arrowheads, so
 *    the path reads dependent -> network -> dependency. The direct service-to-service
 *    edge stays in the model, marked with the networks the pair shares, and is only
 *    rendered when they share none — see `GraphEdge.via` and `showsDirectDependency`.
 *    Two sources, drawn by one rule: a compose `depends_on`, which can only name a
 *    service in the same project, and a `.labview` sidecar's `depends_on`, which is how
 *    a cross-stack dependency can be stated at all. A declared edge carries
 *    `declaredBy`, so no view can pass the operator's statement off as a measurement.
 *
 * **Sharing a network is not a dependency.** Two services on one network can reach each
 * other, and that is the whole of what the membership edges say: they attach a service to
 * the network node and carry no arrowhead. Every fleet has a proxy or monitoring network
 * that half the services are on, and treating co-membership as a relation would draw a
 * line between every pair of them — a statement about the fleet that is simply not true.
 *  - service <-> shared volumes (named volumes, and bind paths used by 2+ stacks)
 *  - ingress/auth hubs, each added only when something observed calls for it:
 *    a Cloudflare tunnel, Traefik, and either the identified SSO provider or a
 *    generic hub when only the mechanism could be established
 *
 * Tunnel ingress is drawn as the path the configuration describes rather than as a
 * straight line to the container. Where a route's origin resolved to another
 * scanned service (see `origins.ts`), that service is drawn as the hop it is:
 * `tunnel -> proxy -> service`. Where the origin resolved to the service itself,
 * or to nothing this scan can see, the direct edge is kept — an unproven hop is
 * never invented.
 *
 * @param proxyKey `${stackId}/${serviceName}` of the proxy whose API was read, when
 * one answered. It is the strongest possible statement about which proxy serves a
 * route — the proxy itself said so — so a router it reported is drawn from that
 * service rather than from the generic hub, in a fleet with no tunnel at all as
 * readily as in one with several.
 * @param nets the fleet's network membership, when the caller already built it. The
 * analyzer does, because the ingress rule needs it first, and sharing it is what makes
 * "shares a network" mean the same thing in the picture as in the classification.
 * @param declared cross-stack dependencies resolved from the fleet's sidecars, from
 * `resolveDeclaredDependencies`. Already resolved and already carrying the networks each
 * pair shares — nothing here re-reads a declaration, so the one place a sidecar reference
 * is interpreted stays the one place its four failure modes are reported.
 */
export function buildGraph(
  stacks: AppStack[],
  proxyKey?: string,
  nets?: NetworkIndex,
  declared: readonly ResolvedDependency[] = [],
): Graph {
  const nodes = new Map<string, GraphNode>();
  const edges: GraphEdge[] = [];
  const addNode = (n: GraphNode) => {
    if (!nodes.has(n.id)) nodes.set(n.id, n);
  };
  // Two routes can describe the same link — a tunnel hostname and the Traefik
  // router serving it both put an edge between the same proxy and service — so
  // edges are kept unique by what they say, not just by id.
  const seenEdge = new Set<string>();
  const addEdge = (e: GraphEdge) => {
    const key = `${e.kind}|${e.source}|${e.target}|${e.label ?? ""}`;
    if (seenEdge.has(key) || seenEdge.has(e.id)) return;
    seenEdge.add(key);
    seenEdge.add(e.id);
    edges.push(e);
  };

  // Hubs (added lazily only if used).
  const hub = { cf: false, traefik: false, authentik: false, auth: false };

  // First pass: count bind-path usage across distinct stacks to find shared data.
  const bindStacks = new Map<string, Set<string>>();
  for (const stack of stacks) {
    for (const svc of stack.services) {
      for (const m of svc.mounts) {
        if (m.type === "bind" && m.source) {
          const key = normalizeBind(m.source);
          if (!bindStacks.has(key)) bindStacks.set(key, new Set());
          bindStacks.get(key)!.add(stack.id);
        }
      }
    }
  }

  const index = nets ?? buildNetworkIndex(stacks);

  /** Every service that exists as a node, for the two hop guards below. */
  const drawn = new Set<string>();
  for (const stack of stacks) {
    for (const svc of stack.services) drawn.add(serviceId(stack, svc));
  }

  /**
   * Which of a service's networks carry a dependency, and which way it runs across
   * them — `out` where this service is the dependent, `in` where something on that
   * network depends on it. This is what lets the two membership edges either side of
   * a network take the arrowheads, so the dependency is drawn *through* the network
   * instead of as a second line beside it.
   *
   * Marked on every network the pair shares, not just one: they can talk over any of
   * them, and picking a favourite would draw an arrow across one network while an
   * equally usable one beside it looked unrelated.
   *
   * Each marking also records whether it came from a compose file or a sidecar, so the
   * arrowhead can be drawn as the kind of statement it is. Nothing marks a flow from
   * membership alone — that is the point of the whole distinction.
   */
  type Flow = { out: boolean; in: boolean; observed: boolean; declared: boolean };
  const flows = new Map<string, Map<string, Flow>>();
  const markFlow = (key: string, net: string, dir: "out" | "in", from: "observed" | "declared") => {
    let byNet = flows.get(key);
    if (!byNet) flows.set(key, (byNet = new Map()));
    const cur = byNet.get(net) ?? { out: false, in: false, observed: false, declared: false };
    cur[dir] = true;
    cur[from] = true;
    byNet.set(net, cur);
  };
  const markPair = (from: string, to: string, source: "observed" | "declared", via?: string[]) => {
    for (const net of via ?? sharedWith(index, from, to)) {
      markFlow(from, net, "out", source);
      markFlow(to, net, "in", source);
    }
  };
  for (const stack of stacks) {
    for (const svc of stack.services) {
      const from = serviceKey(stack, svc);
      for (const dep of svc.dependsOn) {
        const target = stack.services.find((s) => s.name === dep);
        if (!target) continue;
        markPair(from, serviceKey(stack, target), "observed");
      }
    }
  }
  // The declared half, over the networks the resolver already found the pair shares.
  for (const dep of declared) markPair(dep.from, dep.to, "declared", dep.via);

  /** Declared dependencies by dependent, so each service's own loop can emit its edges. */
  const declaredFrom = new Map<string, ResolvedDependency[]>();
  for (const dep of declared) {
    const list = declaredFrom.get(dep.from);
    if (list) list.push(dep);
    else declaredFrom.set(dep.from, [dep]);
  }

  /** Hostnames whose tunnel route terminates at a given hop, for one edge per hop. */
  const viaHop = new Map<string, Set<string>>();

  /** Service nodes shown to be acting as a proxy, tunnel or no tunnel. */
  const proxyRole = new Set<string>();

  for (const stack of stacks) {
    for (const svc of stack.services) {
      const sid = serviceId(stack, svc);
      addNode({
        id: sid,
        label: svc.name,
        kind: "service",
        stack: stack.id,
        auth: svc.auth.method,
        // A node has one background colour, so it carries the most exposed kind. The
        // full set is on the badges beside it — see `GraphNode.ingress`.
        ingress: primaryIngress(svc.ingress),
        running: svc.docker?.running,
      });

      // Networks, under their real docker names, so one `external:` network shared by
      // six stacks is one node with six spokes rather than six unrelated nodes. The
      // node carries what it joins — scope, services, stacks — because that is the
      // whole of what a reader wants from it and counting spokes is not an answer.
      const key = serviceKey(stack, svc);
      for (const name of index.byService.get(key) ?? []) {
        const net = index.byName.get(name);
        if (!net) continue;
        const nid = `net:${name}`;
        addNode({
          id: nid,
          label: name,
          kind: "network",
          scope: net.scope,
          memberCount: net.members.length,
          stackCount: net.stacks.length,
        });
        const flow = flows.get(key)?.get(name);
        addEdge({
          id: `${sid}->${nid}`,
          source: sid,
          target: nid,
          kind: "network",
          ...(flow
            ? {
                flow: flow.out && flow.in ? "both" : flow.out ? "to-network" : "to-service",
                flowSource:
                  flow.observed && flow.declared ? "both" : flow.observed ? "observed" : "declared",
              }
            : {}),
        });
      }

      // depends_on within the stack — all it can name. The edge records which networks
      // the pair shares rather than being dropped when they share one: the relation is
      // real and the drawer names the exact pair from it, while the fleet view draws it
      // through the network instead (`showsDirectDependency`). An empty `via` is the one
      // case with no network to draw it through, and the one worth an arrow of its own.
      for (const dep of svc.dependsOn) {
        const target = stack.services.find((s) => s.name === dep);
        if (target) {
          addEdge({
            id: `${sid}=>${serviceId(stack, target)}`,
            source: sid,
            target: serviceId(stack, target),
            kind: "depends_on",
            via: sharedWith(index, key, serviceKey(stack, target)),
          });
        }
      }

      // The same relation, declared instead of observed — the one way a dependency can
      // cross stacks. Same kind, same `via`, so every rule about how a dependency is
      // drawn applies to it unchanged; `declaredBy` is what keeps the two tellable apart.
      // A pair that is both declared *and* in the compose file collapses to the observed
      // edge above, by the same de-duplication that collapses two routes into one link:
      // the scan saw it, so it need not be taken on anyone's word.
      for (const dep of declaredFrom.get(key) ?? []) {
        addEdge({
          id: `${sid}=>svc:${dep.to}`,
          source: sid,
          target: `svc:${dep.to}`,
          kind: "depends_on",
          via: dep.via,
          declaredBy: { file: dep.file, ...(dep.detail ? { detail: dep.detail } : {}) },
        });
      }

      // Volumes: named volumes always; bind paths only when shared across stacks.
      for (const m of svc.mounts) {
        if (!m.source) continue;
        if (m.type === "volume") {
          const decl = stack.declaredVolumes.find((v) => v.name === m.source);
          const vid = decl?.external ? `vol:ext:${decl.name}` : `vol:${stack.projectName}:${m.source}`;
          addNode({ id: vid, label: m.source, kind: "volume" });
          addEdge({ id: `${sid}~${vid}`, source: sid, target: vid, kind: "volume", label: m.target });
        } else if (m.type === "bind") {
          const key = normalizeBind(m.source);
          if ((bindStacks.get(key)?.size ?? 0) >= 2) {
            const vid = `bind:${key}`;
            addNode({ id: vid, label: m.source, kind: "volume" });
            addEdge({ id: `${sid}~${vid}`, source: sid, target: vid, kind: "volume", label: m.target });
          }
        }
      }

      // Ingress. A tunnel route is drawn through its resolved hop when the origin
      // named another scanned service, and straight at this service otherwise.
      let hopId: string | undefined;
      for (const r of svc.cloudflare) {
        hub.cf = true;
        const label = r.hostname || "tunnel";
        const hop = resolvedHop(r.origin?.hopKey, sid, drawn);
        if (hop) {
          hopId ??= hop;
          // Leg 1 is aggregated per hop below: a fleet routes many hostnames
          // through one proxy, and one edge per route would bundle dozens of
          // parallel lines between the same two nodes.
          if (!viaHop.has(hop)) viaHop.set(hop, new Set());
          viaHop.get(hop)!.add(label);
          addEdge({
            id: `${hop}->${sid}:${r.hostname}`,
            source: hop,
            target: sid,
            kind: "ingress",
            label,
          });
        } else {
          addEdge({
            id: `cf->${sid}:${r.hostname}`,
            source: "ext:cloudflare",
            target: sid,
            kind: "ingress",
            label,
          });
        }
      }
      // Routers the proxy's own API reported for this service. Drawn from that proxy
      // because it is the proxy stating it serves them — no inference involved — and
      // drawn for routers no label declares, which is how ingress configured outside
      // the scanned stacks becomes visible at all. A router the proxy is not serving
      // is left out: an edge is a request path, and that one carries no requests.
      const proxyId = liveProxy(proxyKey, sid, drawn);
      const serving = (svc.traefikLive ?? []).filter(routerIsServing);
      if (proxyId && serving.length) {
        proxyRole.add(proxyId);
        for (const r of serving) {
          addEdge({
            id: `live->${sid}:${r.router}@${r.provider}`,
            source: proxyId,
            target: sid,
            kind: "ingress",
            label: r.hosts[0] ?? r.router,
          });
        }
      }

      // Proxy routes from labels. Those the live read already accounted for are drawn
      // above from the proxy that confirmed them; the rest start at the hop this
      // service's own tunnel origin resolved to, since that service is an observed
      // reverse proxy in front of it. Without either, the responsible proxy is unknown
      // — a Traefik instance picks up routes from the docker socket, which no label
      // records — so the generic hub stands in rather than attributing the route to a
      // proxy found elsewhere.
      const confirmed = new Set(proxyId ? serving.map((r) => r.router.toLowerCase()) : []);
      const routerSource = hopId ?? "ext:traefik";
      for (const r of svc.traefik) {
        if (confirmed.has(r.router.toLowerCase())) continue;
        if (!hopId) hub.traefik = true;
        addEdge({
          id: `tr->${sid}:${r.router}`,
          source: routerSource,
          target: sid,
          kind: "ingress",
          label: r.hosts[0] ?? r.router,
        });
      }
      // Auth hub. Only a service whose provider was actually identified hangs off
      // the named hub; a mechanism-only detection gets the generic one, so the
      // graph never draws a vendor it could not establish from the config.
      if (svc.auth.method.startsWith("authentik")) {
        hub.authentik = true;
        addEdge({
          id: `ak->${sid}`,
          source: "ext:authentik",
          target: sid,
          kind: "auth",
          label: svc.auth.method.replace("authentik-", ""),
        });
      } else if (svc.auth.method === "forward-auth" || svc.auth.method === "other-oauth") {
        hub.auth = true;
        addEdge({
          id: `auth->${sid}`,
          source: "ext:auth",
          target: sid,
          kind: "auth",
          label: svc.auth.method,
        });
      }
    }
  }

  // Leg 1 of every chain: one edge per hop, not per route. Its label carries the
  // hostname when a single route uses the hop, and the count when many do — the
  // hostname-to-service mapping stays visible on leg 2 either way.
  for (const [hop, hostnames] of viaHop) {
    const only = hostnames.size === 1 ? [...hostnames][0] : undefined;
    addEdge({
      id: `cf->${hop}`,
      source: "ext:cloudflare",
      target: hop,
      kind: "ingress",
      label: only ?? `${hostnames.size} hostnames`,
    });
    // The node already exists as an ordinary service; being a hop is an extra fact
    // about it, not a different kind of thing.
    const node = nodes.get(hop);
    if (node) node.role = "proxy";
  }
  // Same fact, established the other way round: a service whose API reported the
  // routes for other services is a proxy whether or not any tunnel lands on it.
  for (const id of proxyRole) {
    const node = nodes.get(id);
    if (node) node.role = "proxy";
  }

  if (hub.cf) addNode({ id: "ext:cloudflare", label: "Cloudflare Tunnel", kind: "external" });
  if (hub.traefik) addNode({ id: "ext:traefik", label: "Traefik", kind: "external" });
  if (hub.authentik) addNode({ id: "ext:authentik", label: "Authentik", kind: "external" });
  if (hub.auth) addNode({ id: "ext:auth", label: "SSO (unidentified)", kind: "external" });

  return { nodes: [...nodes.values()], edges };
}

export function serviceId(stack: AppStack, svc: Service): string {
  return graphServiceId(stack.id, svc.name);
}

/**
 * The graph node for a resolved tunnel hop.
 *
 * `origins.ts` has already established that the hop exists and can reach this
 * service over a shared network, so this only maps its key to a node id and
 * guards the degenerate cases. Without a hop the caller keeps the direct edge.
 */
function resolvedHop(
  hopKey: string | undefined,
  sid: string,
  drawn: ReadonlySet<string>,
): string | undefined {
  if (!hopKey) return undefined;
  const hop = `svc:${hopKey}`;
  if (hop === sid || !drawn.has(hop)) return undefined;
  return hop;
}

/**
 * The graph node for the proxy whose API answered.
 *
 * Guards the two degenerate cases: no proxy was identified, and the proxy is the very
 * service being drawn — a Traefik instance routinely serves a router for its own
 * dashboard, and an edge from a node to itself states nothing.
 */
function liveProxy(
  proxyKey: string | undefined,
  sid: string,
  drawn: ReadonlySet<string>,
): string | undefined {
  if (!proxyKey) return undefined;
  const id = `svc:${proxyKey}`;
  if (id === sid || !drawn.has(id)) return undefined;
  return id;
}

function normalizeBind(p: string): string {
  return p.replace(/\/+$/, "");
}

File: labview/internal/fleet/graph.go
package fleet

import (
	"sort"

	"github.com/nrosier/labview/internal/payload"
)

// Node id prefixes. Every id is `<prefix>:<name>`, which keeps the four diagrams' node sets
// disjoint without a second field to disambiguate them, and makes an edge readable on its own.
const (
	PrefixService  = "svc:"
	PrefixNetwork  = "net:"
	PrefixVolume   = "vol:"
	PrefixHostname = "host:"
	PrefixApp      = "app:"
	PrefixProvider = "provider:"
	PrefixOutpost  = "outpost:"
	PrefixRouter   = "router:"
	// NodeOutside is the single node standing for everything beyond the fleet — the left-hand
	// end of every ingress path (§22.5).
	NodeOutside = "outside"
)

// GraphInput is everything the graph is built from. It is a struct rather than a long parameter
// list because the graph is stage 13 and reads the conclusions of every stage before it; a new
// enrichment adds a field here rather than a second graph.
type GraphInput struct {
	Stacks   []payload.AppStack
	Index    *Index
	Networks *Networks
	Deps     Deps
	// Proxies are the services something resolved to as a hop, or whose proxy API answered.
	// They stay ordinary service nodes; the role only lets the UI colour them (§9).
	Proxies map[string]bool
	// Gates maps a service key to the key of the scanned service that gates it — the outpost
	// whose forward-auth address a middleware on its path names. It is what lets the gate be
	// drawn *on* an ingress path rather than only beside its far end (§22.5).
	Gates map[string]string
	// Authentik and Traefik are optional and used only for their unmatched records, which are
	// shown as unattached nodes rather than hidden (§22.5).
	Authentik *payload.AuthentikSummary
	Traefik   *payload.TraefikSummary
}

// BuildGraph is the one graph object every view reads (§22.4): the whole relation set, uncapped.
// Caps are presentation defaults and belong to a renderer, so nothing is dropped here (§16).
//
// Nodes come out grouped — services in scan order, then drawn networks in scan order, then
// volumes, then everything outside the fleet — and edges grouped by kind in the order of the
// closed set. Two runs over one fleet therefore produce the same bytes (I7).
func BuildGraph(in GraphInput) payload.Graph {
	b := &builder{in: in, seen: map[string]bool{}}

	b.serviceNodes()
	b.networkNodes()
	b.volumeNodes()

	b.membershipEdges()
	b.dependencyEdges()
	b.volumeEdges()
	b.ingressEdges()
	b.authEdges()

	return payload.Graph{Nodes: b.nodes, Edges: b.edges}
}

type builder struct {
	in    GraphInput
	nodes []payload.GraphNode
	edges []payload.GraphEdge
	seen  map[string]bool
}

// node adds a node once. Repeats are normal — two routes reach one service, two stacks mount one
// external volume — and the first spelling wins so the order stays scan order.
func (b *builder) node(n payload.GraphNode) {
	if b.seen[n.ID] {
		return
	}
	b.seen[n.ID] = true
	b.nodes = append(b.nodes, n)
}

// edge adds an edge, deduplicated by id.
func (b *builder) edge(e payload.GraphEdge) {
	if b.seen[e.ID] {
		return
	}
	b.seen[e.ID] = true
	b.edges = append(b.edges, e)
}

func edgeID(kind payload.GraphEdgeKind, source, target string) string {
	return string(kind) + "|" + source + "|" + target
}

// pathID is edgeID for the ingress diagram, which is **one path per route** (§22.5) rather than one
// edge per pair. A hostname served by both a tunnel and a Traefik router reaches its service by two
// mechanisms, and an id built from the pair alone would merge them into one edge labelled with
// whichever was read first — losing a route the payload plainly contains.
func pathID(source, target, label string) string {
	return edgeID(payload.EdgeIngress, source, target) + "|" + label
}

func (b *builder) serviceNodes() {
	for _, stack := range b.in.Stacks {
		for i := range stack.Services {
			svc := stack.Services[i]
			key := Key(stack.ID, svc.Name)
			n := payload.GraphNode{
				ID:      PrefixService + key,
				Label:   svc.Name,
				Kind:    payload.NodeService,
				Stack:   stack.ID,
				Auth:    svc.Auth.Method,
				Ingress: Winner(svc.Ingress),
			}
			// Running is absent rather than false when Docker was not read: a stopped service
			// and a service nothing is known about are different facts (§22.8).
			if svc.Docker != nil {
				running := svc.Docker.Running
				n.Running = &running
			}
			if b.in.Proxies[key] {
				n.Role = payload.RoleProxy
			}
			b.node(n)
		}
	}
}

// networkNodes draws every network except the solo-local ones, which connect nothing. That
// omission is exactly what makes `drawn network nodes + solo-local networks = total networks` a
// checkable identity (§8).
//
// MemberCount and StackCount are counted on the node from the membership index and never
// inferred from the spokes beside it, which a renderer may cap (I1).
func (b *builder) networkNodes() {
	for _, net := range b.in.Networks.All() {
		if !net.Drawn() {
			continue
		}
		members, stacks := len(net.Members), len(net.Stacks)
		b.node(payload.GraphNode{
			ID:          PrefixNetwork + net.Name,
			Label:       net.Name,
			Kind:        payload.NodeNetwork,
			Scope:       net.Scope(),
			MemberCount: &members,
			StackCount:  &stacks,
		})
	}
}

func (b *builder) volumeNodes() {
	for _, stack := range b.in.Stacks {
		for _, vol := range stack.DeclaredVolumes {
			b.node(payload.GraphNode{
				ID:    PrefixVolume + vol.Name,
				Label: vol.Name,
				Kind:  payload.NodeVolume,
			})
		}
	}
}

// membershipEdges is one edge per service per drawn network, carrying the arrowhead and the
// styling of §8.
func (b *builder) membershipEdges() {
	for _, key := range b.in.Index.Keys() {
		for _, name := range b.in.Networks.Of(key) {
			net, ok := b.in.Networks.Get(name)
			if !ok || !net.Drawn() {
				continue
			}
			flow, source := b.leg(key, name)
			b.edge(payload.GraphEdge{
				ID:         edgeID(payload.EdgeNetwork, PrefixService+key, PrefixNetwork+name),
				Source:     PrefixService + key,
				Target:     PrefixNetwork + name,
				Kind:       payload.EdgeNetwork,
				Flow:       flow,
				FlowSource: source,
			})
		}
	}
}

// leg decides what crosses one service's connection to one network.
//
// A dependency contributes only when that network is among the ones the pair shares, so a
// dependency with an empty `via` puts an arrowhead on nothing — which is why it keeps its direct
// edge instead (§8).
func (b *builder) leg(key, network string) (payload.EdgeFlow, payload.EdgeFlowSource) {
	var out, in bool
	var observed, declared bool

	for _, d := range b.in.Deps.Resolved {
		if !contains(d.Via, network) {
			continue
		}
		switch {
		case d.From == key:
			out = true
		case d.To == key:
			in = true
		default:
			continue
		}
		if d.Observed {
			observed = true
		}
		if d.Declared {
			declared = true
		}
	}

	var flow payload.EdgeFlow
	switch {
	case out && in:
		flow = payload.FlowBoth
	case out:
		flow = payload.FlowToNetwork
	case in:
		flow = payload.FlowToService
	default:
		// The common case: the service is on the network and nothing crosses it. Both fields
		// stay absent rather than carrying a member that would read as a claim.
		return "", ""
	}

	// A leg every one of whose dependencies was declared renders dashed; mixed provenance stays
	// solid, because something crossing it was measured.
	switch {
	case observed && declared:
		return flow, payload.FlowSourceBoth
	case declared:
		return flow, payload.FlowSourceDeclared
	default:
		return flow, payload.FlowSourceObserved
	}
}

// dependencyEdges emits every resolved dependency. They are all in the payload: whether a
// renderer draws one is decided by `via`, and that is a rendering rule rather than an analysis
// one (§8, §16).
func (b *builder) dependencyEdges() {
	for _, d := range b.in.Deps.Resolved {
		e := payload.GraphEdge{
			ID:     edgeID(payload.EdgeDependsOn, PrefixService+d.From, PrefixService+d.To),
			Source: PrefixService + d.From,
			Target: PrefixService + d.To,
			Kind:   payload.EdgeDependsOn,
			Label:  labelOf(d.Via),
			Via:    d.Via,
		}
		// declaredBy is present only when a declaration is the *only* account of this pair. A
		// pair compose already resolved is one edge and an observed one, and dashing it would
		// claim the relation was never measured (§14).
		if d.DeclaredOnly() {
			e.DeclaredBy = &payload.EdgeDeclaredBy{File: d.File, Detail: d.Detail}
		}
		b.edge(e)
	}
}

// labelOf names the networks a dependency crosses, which is what the dependency diagram labels
// its edges with (§22.5).
func labelOf(via []string) string {
	switch len(via) {
	case 0:
		return ""
	case 1:
		return via[0]
	default:
		out := via[0]
		for _, name := range via[1:] {
			out += ", " + name
		}
		return out
	}
}

// volumeEdges joins a service to a volume its stack declares. The join applies the same naming
// rule the scan applied — an `external:` or `name:`-overridden volume is verbatim, everything
// else is `${project}_${key}` — in reverse, within the declaring stack. A mount naming no
// declared volume is an anonymous or ad-hoc one and gets no node; it is still on the service.
func (b *builder) volumeEdges() {
	for _, stack := range b.in.Stacks {
		for _, svc := range stack.Services {
			key := Key(stack.ID, svc.Name)
			for _, m := range svc.Mounts {
				if m.Type != payload.MountVolume || m.Source == "" {
					continue
				}
				name, ok := declaredVolume(stack, m.Source)
				if !ok {
					continue
				}
				b.edge(payload.GraphEdge{
					ID:     edgeID(payload.EdgeVolume, PrefixService+key, PrefixVolume+name),
					Source: PrefixService + key,
					Target: PrefixVolume + name,
					Kind:   payload.EdgeVolume,
					Label:  m.Target,
				})
			}
		}
	}
}

func declaredVolume(stack payload.AppStack, source string) (string, bool) {
	for _, vol := range stack.DeclaredVolumes {
		if vol.Name == source || vol.Name == stack.ProjectName+"_"+source {
			return vol.Name, true
		}
	}
	return "", false
}

// ingressEdges draws one path per route: outside → hostname → tunnel or router → origin →
// service (§22.5).
//
// A tunnel whose origin resolved to another scanned service is drawn as a chain through that
// hop, which is §9's table. Every other kind — including `unresolved` — keeps the direct edge:
// an invented hop would be a claim about the path, and dropping the edge would hide a route that
// exists.
func (b *builder) ingressEdges() {
	for _, key := range b.in.Index.Keys() {
		svc := b.in.Index.Service(key)

		for _, route := range svc.Cloudflare {
			if route.Hostname == "" {
				continue
			}
			b.hostname(route.Hostname)
			target := PrefixService + key
			source := PrefixHostname + route.Hostname
			if route.Origin != nil && route.Origin.Kind == payload.OriginFleetService && route.Origin.HopKey != "" {
				hop := PrefixService + route.Origin.HopKey
				b.ingress(source, hop, "tunnel")
				b.ingress(hop, target, route.Origin.Address)
				continue
			}
			b.ingress(source, target, "tunnel")
		}

		for _, route := range svc.Traefik {
			if len(route.Hosts) == 0 {
				continue
			}
			for _, host := range route.Hosts {
				b.hostname(host)
				b.ingress(PrefixHostname+host, PrefixService+key, route.Router)
			}
		}
	}
}

// hostname adds the left-hand end of an ingress path. The `outside → hostname` edge is one per
// hostname rather than one per route — two routes answering for one name enter the fleet at the same
// place — so it keeps the pair-based id and collapses.
func (b *builder) hostname(host string) {
	b.node(payload.GraphNode{ID: NodeOutside, Label: "outside", Kind: payload.NodeExternal})
	b.node(payload.GraphNode{
		ID:    PrefixHostname + host,
		Label: host,
		Kind:  payload.NodeExternal,
	})
	b.edge(payload.GraphEdge{
		ID:     edgeID(payload.EdgeIngress, NodeOutside, PrefixHostname+host),
		Source: NodeOutside,
		Target: PrefixHostname + host,
		Kind:   payload.EdgeIngress,
	})
}

func (b *builder) ingress(source, target, label string) {
	b.edge(payload.GraphEdge{
		ID:     pathID(source, target, label),
		Source: source,
		Target: target,
		Kind:   payload.EdgeIngress,
		Label:  label,
	})
}

// authEdges is the identity-and-auth diagram: which application protects which service and at
// what strength, the providers and outposts behind it, the gate on each path, and every
// unmatched record as an unattached node rather than a hidden one (§22.5).
func (b *builder) authEdges() {
	// The gate first, so a path's gate is drawn even where the identity provider was never read.
	for _, key := range sortedKeys(b.in.Gates) {
		gate := b.in.Gates[key]
		if gate == "" || gate == key || !b.in.Index.Has(gate) || !b.in.Index.Has(key) {
			continue
		}
		b.edge(payload.GraphEdge{
			ID:     edgeID(payload.EdgeAuth, PrefixService+gate, PrefixService+key),
			Source: PrefixService + gate,
			Target: PrefixService + key,
			Kind:   payload.EdgeAuth,
			Label:  string(b.in.Index.Service(key).Auth.Method),
		})
	}

	for _, key := range b.in.Index.Keys() {
		match := b.in.Index.Service(key).Authentik
		if match == nil {
			continue
		}
		for i, app := range match.Applications {
			b.application(app)
			// A strength shorter than the application list reads as `name` for the remainder,
			// never as the strongest (§4.3).
			strength := payload.StrengthName
			if i < len(match.Strength) {
				strength = match.Strength[i]
			}
			b.edge(payload.GraphEdge{
				ID:     edgeID(payload.EdgeAuth, PrefixApp+app.Slug, PrefixService+key),
				Source: PrefixApp + app.Slug,
				Target: PrefixService + key,
				Kind:   payload.EdgeAuth,
				Label:  string(strength),
			})
		}
	}

	if b.in.Authentik != nil {
		for _, un := range b.in.Authentik.UnmatchedApplications {
			b.application(un.Application)
		}
	}
	if b.in.Traefik != nil {
		for _, un := range b.in.Traefik.UnmatchedRouters {
			b.node(payload.GraphNode{
				ID:    PrefixRouter + un.Router.Router + "@" + un.Router.Provider,
				Label: un.Router.Router,
				Kind:  payload.NodeExternal,
			})
		}
	}
}

// application adds one application and the providers and outposts behind it.
func (b *builder) application(app payload.AuthentikApplication) {
	b.node(payload.GraphNode{
		ID:    PrefixApp + app.Slug,
		Label: app.Name,
		Kind:  payload.NodeExternal,
	})
	for _, provider := range app.Providers {
		b.node(payload.GraphNode{
			ID:    PrefixProvider + provider.Name,
			Label: provider.Name,
			Kind:  payload.NodeExternal,
		})
		b.edge(payload.GraphEdge{
			ID:     edgeID(payload.EdgeAuth, PrefixProvider+provider.Name, PrefixApp+app.Slug),
			Source: PrefixProvider + provider.Name,
			Target: PrefixApp + app.Slug,
			Kind:   payload.EdgeAuth,
			Label:  string(provider.Kind),
		})
		// Outposts come off the API in whatever order it listed them; sorting them here is what
		// keeps two reads of one provider producing the same graph (I7).
		outposts := append([]string(nil), provider.Outposts...)
		sort.Strings(outposts)
		for _, outpost := range outposts {
			b.node(payload.GraphNode{
				ID:    PrefixOutpost + outpost,
				Label: outpost,
				Kind:  payload.NodeExternal,
			})
			b.edge(payload.GraphEdge{
				ID:     edgeID(payload.EdgeAuth, PrefixOutpost+outpost, PrefixProvider+provider.Name),
				Source: PrefixOutpost + outpost,
				Target: PrefixProvider + provider.Name,
				Kind:   payload.EdgeAuth,
			})
		}
	}
}
